Implement API AppendAllBytes

MojtabaTajik opened the pull request:
Added AppendAllBytes API and the async version with the corresponding tests.

Fix #84532

Commit:
Implement AppendAllBytes API

Commit:
Add AppendAllBytes declaration

Commit:
Add unit tests for AppendAllBytes

Commit:
Add AppendAllBytesAsync tests

Commit:
Add AppendAllBytesAsync implemntation

Commit:
Add AppendAllBytesAsync definition

MojtabaTajik committed:
Removed unnecessary seek capability

Commit:
Changed FileMode to Append to make it more readable and match the behavior

MojtabaTajik committed:
Merge branch 'dotnet:main' into API_AppendAllBytes-84532

MojtabaTajik committed:
Merge branch 'dotnet:main' into API_AppendAllBytes-84532

Commit:
Move AppendAllBytes to separate class

Commit:
Add extra tests to cover more scenarios

Commit:
Add tests to cover AppendAllBytesAsync

Commit:
Remove unnecessary delete test file, becuase its removed by cleanup

MojtabaTajik committed:
Merge branch 'dotnet:main' into API_AppendAllBytes-84532

MojtabaTajik committed:
Add API documentation

MojtabaTajik committed:
Merge branch 'API_AppendAllBytes-84532' of https://github.com/MojtabaTajik/runtime into API_AppendAllBytes-84532

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
         }
 
+        /// <summary>
+        /// Appends the specified byte array to the end of the file at the given path. 
```
```suggestion
        /// Appends the specified byte array to the end of the file at the given path.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// <param name="path">The path of the file to which the bytes should be appended.</param>
+        /// <param name="bytes">The byte array to append to the file.</param>
+        /// <exception cref="T:System.ArgumentException">
+        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters 
```
```suggestion
        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// If the file doesn't exist, this method creates a new file.
+        /// </summary>
+        /// <param name="path">The path of the file to which the bytes should be appended.</param>
+        /// <param name="bytes">The byte array to append to the file.</param>
```
```suggestion
        /// <param name="bytes">The bytes to append to the file.</param>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// Appends the specified byte array to the end of the file at the given path. 
+        /// If the file doesn't exist, this method creates a new file.
+        /// </summary>
+        /// <param name="path">The path of the file to which the bytes should be appended.</param>
```
Follow existing conventions.
```suggestion
        /// <param name="path">The file to append to.</param>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// <param name="path">The path of the file to which the bytes should be appended.</param>
+        /// <param name="bytes">The byte array to append to the file.</param>
+        /// <exception cref="T:System.ArgumentException">
+        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters 
+        /// (which can be queried using the <see cref="M:System.IO.Path.GetInvalidPathChars" /> method), 
+        /// or when <paramref name="bytes" /> is null or empty.
```
```suggestion
        /// <paramref name="path" /> is a zero-length string, contains invalid characters
        /// (which can be queried using the <see cref="M:System.IO.Path.GetInvalidPathChars" /> method),
        /// or when <paramref name="bytes" /> is null or empty.
```

Jozkee replied:
> when <paramref name="bytes" /> is null or empty

We should not throw when `bytes` is empty. I think this is just a documentation error, no?

MojtabaTajik replied:
Indeed, I need to correct the documentation. The behaviour is to throw an exception for a null param, not an empty byte array.

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// <param name="cancellationToken">A token to monitor for cancellation requests. The default value is <see cref="T:System.Threading.CancellationToken.None" />.</param>
+        /// <returns>A task that represents the asynchronous append operation.</returns>
+        /// <exception cref="T:System.ArgumentException">
+        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters 
```
```suggestion
        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,63 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// <returns>A task that represents the asynchronous append operation.</returns>
+        /// <exception cref="T:System.ArgumentException">
+        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters 
+        /// (which can be queried using the <see cref="M:System.IO.Path.GetInvalidPathChars" /> method), 
```
```suggestion
        /// (which can be queried using the <see cref="M:System.IO.Path.GetInvalidPathChars" /> method),
```

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytes.cs:
```diff
@@ -0,0 +1,103 @@
+        [Fact]
+        public void InvalidParameters()
+        {
+            Assert.Throws<ArgumentException>(() => File.AppendAllBytes(string.Empty, new byte[0]));
```
Is `new byte[0]` an invalid parameter?

MojtabaTajik replied:
Nope, the `string.Empty` for the path parameter. We have same tests for other APIs, for example [File_ReadWriteAllBytes](https://github.com/dotnet/runtime/blob/10ce9fb494c9f2a16a757e1c8a1f902c28be4381/src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs#L25).

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytes.cs:
```diff
@@ -0,0 +1,103 @@
+            using (File.Create(path))
+            {
+                Assert.Throws<IOException>(() => File.AppendAllBytes(path, bytes));
+                Assert.Throws<IOException>(() => File.ReadAllBytes(path));
```
This is already tested in https://github.com/dotnet/runtime/blob/00f1e0c94a3853f65d153179c1981d5b3f01c5e6/src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs#L79
```suggestion
```

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytesAsync.cs:
```diff
@@ -0,0 +1,104 @@
+            using (File.Create(path))
+            {
+                await Assert.ThrowsAsync<IOException>(async () => await File.AppendAllBytesAsync(path, bytes));
+                await Assert.ThrowsAsync<IOException>(async () => await File.ReadAllBytesAsync(path));
```
```suggestion
```

MojtabaTajik committed:
Remove extra space

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Remove extra space

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
update description of API

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Update param description

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Remove extra space

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Remove extra space

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Remove unnecessary test scenario

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Remove unnecessary test scenario

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Remove extra space

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Fix docusmentation of API

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytes.cs:
```diff
@@ -0,0 +1,102 @@
+                File.SetAttributes(path, FileAttributes.Normal);
+            }
+        }
+
```
Whitespace
```suggestion
```

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytes.cs:
```diff
@@ -0,0 +1,102 @@
+        public void EmptyContentCreatesFile()
+        {
+            string path = GetTestFilePath();
+            File.AppendAllBytes(path, new byte[0]);
+            Assert.True(File.Exists(path));
```
```suggestion
            Assert.False(File.Exists(path));
            File.AppendAllBytes(path, new byte[0]);
            Assert.True(File.Exists(path));
```

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytes.cs:
```diff
@@ -0,0 +1,102 @@
+        public void InvalidParameters()
+        {
+            Assert.Throws<ArgumentException>(() => File.AppendAllBytes(string.Empty, new byte[0]));
+        }
```
Can you please add a test that throws if the directory does not exist (for async too) e.g:
https://github.com/dotnet/runtime/blob/10ce9fb494c9f2a16a757e1c8a1f902c28be4381/src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllText.cs#L42-L46

MojtabaTajik replied:
Sure 👌

MojtabaTajik replied:
Done.

Jozkee commented on src/libraries/System.IO.FileSystem/tests/File/AppendAllBytesAsync.cs:
```diff
@@ -0,0 +1,103 @@
+                File.SetAttributes(path, FileAttributes.Normal);
+            }
+        }
+
+    }
+
```
```suggestion
    }
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// </summary>
+        /// <param name="path">The file to append to.</param>
+        /// <param name="bytes">The bytes to append to the file.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <paramref name="path"/> is a zero-length string, contains only white space,
+        /// or contains invalid characters as defined by <see cref="System.IO.Path.GetInvalidPathChars"/>.
+        /// </exception>
```
Don't use "Throw when".
```suggestion
        /// <exception cref="System.ArgumentException">
        /// <paramref name="path"/> is a zero-length string, contains only white space, or contains one more invalid characters defined by the <see cref="System.IO.Path.GetInvalidPathChars"/> method.
        /// </exception>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// or contains invalid characters as defined by <see cref="System.IO.Path.GetInvalidPathChars"/>.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when <paramref name="path"/> is null, or when <paramref name="bytes"/> is null.
```
```suggestion
        /// Either <paramref name="path"/> or <paramref name="bytes"/> is null.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// Asynchronously appends the specified byte array to the end of the file at the given path.
+        /// If the file doesn't exist, this method creates a new file. If the operation is canceled, the task will be returned in a canceled state.
+        /// </summary>
+        /// <param name="path">The path of the file to which the bytes should be appended.</param>
```
```suggestion
        /// <param name="path">The file to append to.</param>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// </summary>
+        /// <param name="path">The path of the file to which the bytes should be appended.</param>
+        /// <param name="bytes">The byte array to append to the file.</param>
+        /// <param name="cancellationToken">A token to monitor for cancellation requests. The default value is <see cref="T:System.Threading.CancellationToken.None" />.</param>
```
```suggestion
        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="T:System.Threading.CancellationToken.None" />.</param>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// <param name="bytes">The byte array to append to the file.</param>
+        /// <param name="cancellationToken">A token to monitor for cancellation requests. The default value is <see cref="T:System.Threading.CancellationToken.None" />.</param>
+        /// <returns>A task that represents the asynchronous append operation.</returns>
+        /// <exception cref="T:System.ArgumentException">
+        /// Thrown when <paramref name="path" /> is a zero-length string, contains invalid characters
+        /// (which can be queried using the <see cref="M:System.IO.Path.GetInvalidPathChars" /> method),
+        /// or when <paramref name="bytes" /> is null or empty.
+        /// </exception>
```
```suggestion
        /// <exception cref="T:System.ArgumentException">
        /// <paramref name="path"/> is a zero-length string, contains only white space, or contains one more invalid characters defined by the <see cref="System.IO.Path.GetInvalidPathChars"/> method.
        /// </exception>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// or when <paramref name="bytes" /> is null or empty.
+        /// </exception>
+        /// <exception cref="T:System.ArgumentNullException">
+        /// Thrown when <paramref name="path" /> is <see langword="null" />.
```
```suggestion
        /// Either <paramref name="path"/> or <paramref name="bytes"/> is null.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// </exception>
+        /// <exception cref="T:System.ArgumentNullException">
+        /// Thrown when <paramref name="path" /> is <see langword="null" />.
+        /// </exception>
```
```suggestion
        /// </exception>
        /// <exception cref="T:System.OperationCanceledException">
        /// The cancellation token was canceled. This exception is stored into the returned task.
        /// </exception>
```

jozkee reviewed (commented):
Almost there.

When you commit changes, can you please push when you have all changes ready or use the "Add suggestion to batch" in GH; each push may trigger a CI run which is more expensive than having one after all feedback is addressed.

MojtabaTajik committed:
Apply suggestions from code review

- Remove unwanted spaces in files
- Fix API documentation
- Add assertation for file exists

Co-authored-by: David Cantú <dacantu@microsoft.com>

MojtabaTajik committed:
Add test for scenario when directory not exists

MojtabaTajik committed:
Fix Async documentation and match it with none-async version

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,60 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// <param name="cancellationToken">A token to monitor for cancellation requests. The default value is <see cref="System.Threading.CancellationToken.None"/>.</param>
+        /// <returns>A task that represents the asynchronous append operation.</returns>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <paramref name="path"/> is a zero-length string, contains only white space,
+        /// or contains invalid characters as defined by <see cref="System.IO.Path.GetInvalidPathChars"/>.
```
```suggestion
        /// <paramref name="path"/> is a zero-length string, contains only white space, or contains one more invalid characters defined by the <see cref="System.IO.Path.GetInvalidPathChars"/> method.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,60 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// or contains invalid characters as defined by <see cref="System.IO.Path.GetInvalidPathChars"/>.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when <paramref name="path"/> is null, or when <paramref name="bytes"/> is null.
```
```suggestion
        /// Either <paramref name="path"/> or <paramref name="bytes"/> is null.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,60 @@ public static void WriteAllBytes(string path, byte[] bytes)
+
+        /// <summary>
+        /// Asynchronously appends the specified byte array to the end of the file at the given path.
+        /// If the file does not exist, a new file will be created. If the operation is canceled, the task will return in a canceled state.
```
```suggestion
        /// If the file doesn't exist, this method creates a new file. If the operation is canceled, the task will return in a canceled state.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,60 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// </summary>
+        /// <param name="path">The file to append to.</param>
+        /// <param name="bytes">The bytes to append to the file.</param>
+        /// <param name="cancellationToken">A token to monitor for cancellation requests. The default value is <see cref="System.Threading.CancellationToken.None"/>.</param>
```
```suggestion
        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="System.Threading.CancellationToken.None"/>.</param>
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,60 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when <paramref name="path"/> is null, or when <paramref name="bytes"/> is null.
+        /// </exception>
```
```suggestion
        /// </exception>
        /// <exception cref="T:System.OperationCanceledException">
	    /// The cancellation token was canceled. This exception is stored into the returned task.
	    /// </exception>
```

jozkee committed:
Apply suggestions from code review

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// Either <paramref name="path"/> or <paramref name="bytes"/> is null.
+        /// </exception>
+        /// <exception cref="T:System.OperationCanceledException">
+	    /// The cancellation token was canceled. This exception is stored into the returned task.
```
```suggestion
        /// The cancellation token was canceled. This exception is stored into the returned task.
```

Jozkee commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -689,6 +689,62 @@ public static void WriteAllBytes(string path, byte[] bytes)
+        /// </exception>
+        /// <exception cref="T:System.OperationCanceledException">
+	    /// The cancellation token was canceled. This exception is stored into the returned task.
+	    /// </exception>
```
```suggestion
        /// </exception>
```

jozkee committed:
Change tabs for spaces
